lib/logstash/outputs/elasticsearch/data_stream_support.rb

```diff
@@ -0,0 +1,249 @@
+      ds_default = Gem::Version.create(LOGSTASH_VERSION) >= Gem::Version.create(DATA_STREAMS_ENABLED_BY_DEFAULT_LS_VERSION)
+
+      return false if @data_stream.nil? && !ds_default # data_stream => ... not set on LS 7.x
+
+      if ds_default # LS 8.0
+        return false unless valid_data_stream_config
+
+        assert_es_version_supports_data_streams(true)
+        @logger.debug 'Configuration is data_stream compatible'
+        return true
+      end
+
+      # LS 7.x data_stream => auto
+      valid_data_stream_config && assert_es_version_supports_data_streams(false)
+    end
+
+    # an {event_action_tuple} replacement when a data-stream configuration is detected
+    def data_stream_event_action_tuple(event)
+      data_stream_event_sync(event) if data_stream_sync_fields
+      ['create', common_event_params(event), event.to_hash] # action always 'create'
+    end
+
+    DATA_STREAM_SYNC_FIELDS = [ 'type', 'dataset', 'namespace' ].freeze
+
+    def data_stream_event_sync(event)
+      data_stream = event.get('data_stream')
+      if data_stream.is_a?(Hash)
+        unless data_stream_auto_routing
+          sync_fields = DATA_STREAM_SYNC_FIELDS.select { |name| data_stream.key?(name) && data_stream[name] != send(:"data_stream_#{name}") }
+          if sync_fields.any? # these fields will need to be overwritten
+            info = sync_fields.inject({}) { |info, name| info[name] = data_stream[name]; info }
+            info[:event] = event.to_hash
+            @logger.warn "Some data_stream fields are out of sync, these will be updated to reflect data-stream name", info
+
+            sync_fields.each { |name| data_stream[name] = nil } # fallback to ||= bellow
+          end
+        end
+      else
+        unless data_stream.nil?
+          @logger.warn "Invalid 'data_stream' field type, due fields sync will overwrite", value: data_stream, event: event.to_hash
+        end
+        data_stream = Hash.new
+      end
+
+      data_stream['type'] ||= data_stream_type
+      data_stream['dataset'] ||= data_stream_dataset
+      data_stream['namespace'] ||= data_stream_namespace
+
+      event.set('data_stream', data_stream)
+    end
+
+    module Validator
+
+      # @override {LogStash::Config::Mixin::validate_value} to provide :ds_identifier_string validator
+      # @param value [Array<Object>]
+      # @param validator [nil,Array,Symbol]
+      # @return [Array(true,Object)]: if validation is a success, a tuple containing `true` and the coerced value
+      # @return [Array(false,String)]: if validation is a failure, a tuple containing `false` and the failure reason.
+      def validate_value(value, validator)
+        if validator == :dataset_identifier
+          return validate_dataset_identifier(value)
+        end
+        if validator == :namespace_identifier
+          return validate_namespace_identifier(value)
+        end
+        return super unless validator == :ds_identifier_string
+
+        super(value, validator)
```

Comment:
This bit is redundant:
~~~
        return super unless validator == :ds_identifier_string

        super(value, validator)
~~~

Addititonally, I think this can be made more clear by using a `case` statement:

```suggestion
        case validator
        when :dataset_identifier   then validate_dataset_identifier(value)
        when :namespace_identifier then validate_namespace_identifier(value)
        else super
        end
```

The comment at the top about which validators it provides needs updating since `:ds_identifier_string` is no longer useful.